src/BSON/Int64.c

```diff
@@ -0,0 +1,377 @@
+ * limitations under the License.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <php.h>
+#include <Zend/zend_interfaces.h>
+#include <ext/standard/php_var.h>
+#if PHP_VERSION_ID >= 70000
+#include <zend_smart_str.h>
+#else
+#include <ext/standard/php_smart_str.h>
+#endif
+
+#include "phongo_compat.h"
+#include "php_phongo.h"
+#include "php_bson.h"
+
+zend_class_entry* php_phongo_int64_ce;
+
+/* Initialize the object and return whether it was successful. */
+static bool php_phongo_int64_init(php_phongo_int64_t* intern, int64_t integer) /* {{{ */
+{
+	intern->integer     = integer;
+	intern->initialized = true;
+
+	return true;
+} /* }}} */
+
+/* Initialize the object from a numeric string and return whether it was
+ * successful. An exception will be thrown on error. */
+static bool php_phongo_int64_init_from_string(php_phongo_int64_t* intern, const char* s_integer, phongo_zpp_char_len s_integer_len TSRMLS_DC) /* {{{ */
+{
+	int64_t integer;
+	char*   endptr = NULL;
+
+	/* bson_ascii_strtoll() sets errno if conversion fails. If conversion
+	 * succeeds, we still want to ensure that the entire string was parsed. */
+
+	integer = bson_ascii_strtoll(s_integer, &endptr, 10);
+
+	if (errno || (endptr && endptr != ((const char*) s_integer + s_integer_len))) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Error parsing \"%s\" as 64-bit integer for %s initialization", s_integer, ZSTR_VAL(php_phongo_int64_ce->name));
+		return false;
+	}
+
+	return php_phongo_int64_init(intern, integer);
+} /* }}} */
+
+/* Initialize the object from a HashTable and return whether it was successful.
+ * An exception will be thrown on error. */
+static bool php_phongo_int64_init_from_hash(php_phongo_int64_t* intern, HashTable* props TSRMLS_DC) /* {{{ */
+{
+#if PHP_VERSION_ID >= 70000
+	zval* value;
+
+	if ((value = zend_hash_str_find(props, "int", sizeof("int") - 1)) && Z_TYPE_P(value) == IS_STRING) {
+		return php_phongo_int64_init_from_string(intern, Z_STRVAL_P(value), Z_STRLEN_P(value) TSRMLS_CC);
+	}
+#else
+	zval** value;
+
+	if (zend_hash_find(props, "int", sizeof("int"), (void**) &value) == SUCCESS && Z_TYPE_PP(value) == IS_STRING) {
+		return php_phongo_int64_init_from_string(intern, Z_STRVAL_PP(value), Z_STRLEN_PP(value) TSRMLS_CC);
+	}
+#endif
+
+	phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "%s initialization requires \"int\" string field", ZSTR_VAL(php_phongo_int64_ce->name));
+	return false;
+} /* }}} */
+
+/* {{{ proto string MongoDB\BSON\Int64::__toString()
+   Return the Int64's value as a string. */
+static PHP_METHOD(Int64, __toString)
+{
+	php_phongo_int64_t* intern;
+	char                s_integer[24];
+	int                 s_integer_len;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	intern = Z_INT64_OBJ_P(getThis());
+
+	s_integer_len = snprintf(s_integer, sizeof(s_integer), "%" PRId64, intern->integer);
```

Comment:
Although I prefer spprintf in general, I guess it makes sense to use a stack buffer here.

Reply:
Aye, this static buffer approach was done in a few places elsewhere. I see no reason we can't do this any time we're creating a string from a 64-bit integer, which is why I also refactored UTCDateTime in #857. That said, if you have the same concerns about appending a null pointer, would you prefer I use `spprintf` in all of these places instead?

Reply:
Note: we agreed to keep use of `snprintf`.